src/LuaMultiThreaded2.cpp

```diff
@@ -67,7 +68,99 @@ extern "C" int errorHandler(LuaState * L)
 
 
 
+/** Creates a new mutex object.
+Returns: The mutex object with the functions in mutexObjFuncs as methods in it's metatable. */
+extern "C" static int mutexNew(LuaState * aState)
+{
+	lua_pushcfunction(aState, errorHandler);
+
+	// Push the (currently empty) mutex object to the Lua side
+	auto mutexObj = reinterpret_cast<std::mutex **>(lua_newuserdata(aState, sizeof(std::mutex **)));
+	luaL_setmetatable(aState, MUTEX_METATABLE_NAME);
+
+	// Create the new mutex:
+	*mutexObj = new std::mutex();
+	return 1;
+}
+
+
+
+
+
+/** Executes the provided function while the mutex is locked. */
+extern "C" static int mutexDoWhileLocked(LuaState * aState)
+{
+	auto mutexObj = reinterpret_cast<std::mutex **>(luaL_checkudata(aState, 1, MUTEX_METATABLE_NAME));
+	if (mutexObj == nullptr)
+	{
+		luaL_argerror(aState, 0, "'mutex' expected");
+		return 0;
+	}
+	if (*mutexObj == nullptr) 
+	{
+		luaL_argerror(aState, 0, "Internal error: 'mutex' object is invalid");
+		return 0;
+	}
+	(*mutexObj)->lock();
+	auto numParams = lua_gettop(aState);
+	luaL_checktype(aState, 2, LUA_TFUNCTION);
+	lua_pcall(aState, numParams - 2, 0, 0);
+	(*mutexObj)->unlock();
+
+	return 0;
+}
+
+
+
+
+
+/** Locks the provided mutex. */
+extern "C" static int mutexLock(LuaState * aState) 
+{
+	auto mutexObj = reinterpret_cast<std::mutex **>(luaL_checkudata(aState, 1, MUTEX_METATABLE_NAME));
+	if (mutexObj == nullptr)
+	{
+		luaL_argerror(aState, 0, "'mutex' expected");
+		return 0;
+	}
+	if (*mutexObj == nullptr)
+	{
+		luaL_argerror(aState, 0, "Internal error: 'mutex' object is invalid");
+		return 0;
+	}
+	(*mutexObj)->lock();
+	return 0;
+}
+
+
+
+
+
+/** Unlocks the provided mutex. */
+extern "C" static int mutexUnlock(LuaState * aState) 
+{
+	auto mutexObj = reinterpret_cast<std::mutex **>(luaL_checkudata(aState, 1, MUTEX_METATABLE_NAME));
+	if (mutexObj == nullptr)
+	{
+		luaL_argerror(aState, 0, "'mutex' expected");
+		return 0;
+	}
+	if (*mutexObj == nullptr)
+	{
+		luaL_argerror(aState, 0, "Internal error: 'mutex' object is invalid");
+		return 0;
+	}
+	(*mutexObj)->unlock();
```

Comment:
Visual Studio is giving a warning here, but apparently it's most likely a bug: https://stackoverflow.com/a/58710813

`Caller failing to hold lock '*(*mutexObj)' before calling function 'std::_Mutex_base::unlock'.`

Reply:
I'm not sure about it being the same bug, but it's definitely a false positive, something that the static analyzer cannot catch.